### PR TITLE
Add haptic feedback interface

### PR DESCRIPTION
**Pull request by rriveramcrus**
Adds a haptic driver API and a DRV2605 haptic device driver.

### CODE REVIEW

File: samples/drivers/drv2605/src/main.c [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ */
+
+#include <errno.h>
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/haptics.h>
+#include <zephyr/drivers/haptics/drv2605.h>
```
Comment (bjarki-andreasen) [inline samples/drivers/drv2605/src/main.c:14]: Since the configuration of the haptic device is vendor/device specific, it is more efficient to move the/all configuration functions to the vendor/device specific header :)

Specifically, the API `haptics_source_config(dev, HAPTICS_SOURCE_ROM, &rom_data);` is not portable, whereas `haptics_start_output(dev);` is.

Forcing the config struct, in this case `struct drv2605_rom_data rom_data = {};` to be passed by void reference through a "generic" API is less safe than creating a specific API for the drv2605 which takes a type safe combination of arguments, like:

```
int haptics_source_config_rom(const struct device *dev, const struct drv2605_rom_data *data);
```
Which is completely type safe (except for the *dev pointer but that's a Zephyr thing).

In summary, device specific configs/features which are not generic enough to be abstractly used by the application (one can switch from one haptic device of one vendor to another, without changing the application), write device specific typesafe APIs for it. Only expose the generic APIs in the haptics API

Comment (rriveramcrus) [reply samples/drivers/drv2605/src/main.c:14]: ack, I see your point here.

File: include/zephyr/drivers/haptics.h [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	HAPTICS_SOURCE_RAM,
+	/* The playback source is an external PWM signal */
+	HAPTICS_SOURCE_PWM,
+	/* The playback source is an external analog signal */
+	HAPTICS_SOURCE_ANALOG,
+	/* The playback source is an external digital signal */
+	HAPTICS_SOURCE_DIGITAL,
+	/* The playback source is an external audio signal */
+	HAPTICS_SOURCE_AUDIO,
+};
```
Comment (bjarki-andreasen) [inline include/zephyr/drivers/haptics.h:39]: This struct is now stale :)

File: dts/bindings/haptics/ti,drv2605.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+description: |
+  DRV2605 Haptic Driver for ERM and LRA with built-in library and smart-loop
+  architecture.
+
+compatible: "ti,drv2605"
+
+include: i2c-device.yaml
+
+properties:
+  ti,actuator-mode:
```
Comment (bjarki-andreasen) [inline dts/bindings/haptics/ti,drv2605.yaml:13]: We typically only use dashes in dts properties, and exclude the vendor prefix, so `ti,actuator-mode` would be `actuator-mode`

File: dts/bindings/haptics/ti,drv2605.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    type: string
+    enum:
+      - "ERM Mode"
+      - "LRA Mode"
+  ti,feedback-brake-factor:
+    type: string
+    enum:
+      - "1x"
```
Comment (bjarki-andreasen) [inline dts/bindings/haptics/ti,drv2605.yaml:21]: Typically string enums are formatted like macros, with no spaces, and shortened as much as possible:
```
- "1X"
- "2X"
...
- "DISABLED"
```
This makes it easy to directly use the value within the drivers by concatenation like:
```
#define TI_FEEDBACK_BRAKE_FACTOR_1X 1
#define TI_FEEDBACK_BRAKE_FACTOR_2X 2
...

uint8_t break_factor = _CONCAT(TI_FEEDBACK_BRAKE_FACTOR_, DT_INST_STRING_TOKEN(inst, ti_feedback_brake_factor))
```

File: drivers/haptics/Kconfig.drv2605 [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+	help
+	  Enable I2C-based driver for the Texas Instruments DRV2605 Haptics
+	  Driver.
+
+config HAPTICS_DRV2605_INIT_PRIORITY
+	int "DRV2605 haptic device driver init priority"
+	default KERNEL_INIT_PRIORITY_DEVICE
+	depends on HAPTICS_DRV2605
+	help
+	  Init priority for the Texas Instruments DRV2605 Haptic device driver.
```
Comment (bjarki-andreasen) [inline drivers/haptics/Kconfig.drv2605:18]: Just use the HAPTICS_INIT_PRIORITY in the device driver :) Additionally, to add driver specific configs, hide them behind
```
if HAPTICS_DRV2605
...
endif
```
rather than `depends on HAPTICS_DRV2605`

File: include/zephyr/drivers/haptics/drv2605.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * @param dev Pointer to the device structure for haptic device instance
+ * @param source The type of haptic signal source desired
+ * @param config_data Pointer to the configuration data union for the source
+ *
+ * @retval 0 if successful
+ * @retval -ENOTSUP if the signal source is not supported
+ * @retval < 0 if failed
+ */
+int drv2605_haptic_config(const struct device *dev, enum haptics_source source,
```
Comment (bjarki-andreasen) [inline include/zephyr/drivers/haptics/drv2605.h:67]: The source struct should be moved here instead of being defined in the `haptic.h` header

File: samples/drivers/drv2605/prj.conf [hunk 7]
```diff
@@ -0,0 +1,3 @@
+CONFIG_LOG=y
+CONFIG_HAPTICS=y
+CONFIG_HAPTICS_DRV2605=y
```
Comment (bjarki-andreasen) [inline samples/drivers/drv2605/prj.conf:3]: The device driver ~should select~ already selects the driver automatically using `depends on DT_HAS_TI_DRV2605_ENABLED` and `default y` making this selection redundant :)

File: include/zephyr/drivers/haptics/drv2605.h (hunk 6 above)
Comment (rriveramcrus) [reply include/zephyr/drivers/haptics/drv2605.h:67]: ack

File: dts/bindings/haptics/ti,drv2605.yaml (hunk 4 above)
Comment (rriveramcrus) [reply dts/bindings/haptics/ti,drv2605.yaml:21]: ack

File: dts/bindings/haptics/ti,drv2605.yaml (hunk 3 above)
Comment (rriveramcrus) [reply dts/bindings/haptics/ti,drv2605.yaml:13]: ack

File: samples/drivers/drv2605/prj.conf (hunk 7 above)
Comment (rriveramcrus) [reply samples/drivers/drv2605/prj.conf:3]: ack

File: drivers/haptics/Kconfig.drv2605 (hunk 5 above)
Comment (rriveramcrus) [reply drivers/haptics/Kconfig.drv2605]: ack, I'll just use `HAPTICS_INIT_PRIORITY` :)

**Review by bjarki-andreasen — DISMISSED** [review]
I think this looks good :)

File: samples/drivers/drv2605/src/main.c [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+#include <zephyr/sys/util.h>
+#include <sys/_stdint.h>
+
+#define LOG_LEVEL 4
+
+LOG_MODULE_REGISTER(main);
+
+static struct drv2605_rom_data rom_data = {
```
Comment (kartben) [inline samples/drivers/drv2605/src/main.c:22]: Would it make sense at all for the driver to have a default ROM? It kind of bothers me that the first sample being introduced for the new subsystem can't be 100% generic, but maybe that's just the way it is 
.. so not directly a request for change, just asking in case there is a way around this :)
Also wondering whether ROM information could be provided as devicetree overlay? (i.e. if it's "ROM" one can probably argue it's very much hardware bound and not really dynamic configuration?)

**Review by kartben — CHANGES_REQUESTED** [review]

File: samples/drivers/drv2605/CMakeLists.txt
[the remark this answers is not in the document]
Comment (rriveramcrus) [reply samples/drivers/drv2605/CMakeLists.txt:1]: ack and ack

File: include/zephyr/drivers/haptics.h (hunk 2 above)
Comment (rriveramcrus) [reply include/zephyr/drivers/haptics.h]: ack

File: samples/drivers/drv2605/src/main.c (hunk 8 above)
Comment (rriveramcrus) [reply samples/drivers/drv2605/src/main.c:22]: Hey @kartben, the ROM refers to the bank of licensed waveforms built-in on the DRV2605. By default the sequencer queue on the DRV2605 is empty, otherwise I would have a generic sample app this chip could make use of where we play the default waveform. So unfortunately I think that's just the way it is.

We could introduce configuring the sequencer and the related controls to a default condition in the devicetree, but I feel that this will be set at runtime most of the time. If you are using ROM as a playback source there are various waveforms in each bank that could be used for different feedback events. Alternatively the other playback sources supported by the DRV2605 rely on some external stimulus (PWM, analog signal, analog audio, streamed values) to be applied at runtime.

**Review by kartben — CHANGES_REQUESTED** [review]
just spotted minor doxygen issue, lgtm otherwise - thanks!

**Review by kartben — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]
Cool!

**Review by fabiobaltieri — COMMENTED** [review]
Can you add a maintainers file entry as well?

**Review by bjarki-andreasen — APPROVED** [review]

**Comment by rriveramcrus** [comment]
@fabiobaltieri,

> Can you add a maintainers file entry as well?

Will do in a subsequent PR later today.

File: drivers/haptics/drv2605.c [hunk 9]
```diff
@@ -0,0 +1,621 @@
...
+
+#define HAPTICS_DRV2605_DEFINE(inst)                                                               \
+                                                                                                   \
+	static const struct drv2605_config drv2605_config_##inst = {                               \
+		.i2c = I2C_DT_SPEC_INST_GET(inst),                                                 \
+		.en_gpio = GPIO_DT_SPEC_INST_GET_OR(inst, ti_en_gpios, {}),                        \
+		.in_trig_gpio = GPIO_DT_SPEC_INST_GET_OR(inst, ti_in_trig_gpios, {}),              \
+		.feedback_brake_factor = DT_INST_ENUM_IDX_OR(inst, ti_feedback_brake_factor, 3),   \
+		.loop_gain = DT_INST_ENUM_IDX_OR(inst, ti_loop_gain, 2),                           \
+		.actuator_mode = DT_INST_ENUM_IDX_OR(inst, ti_actuator_mode, 0),                   \
```
Comment (kartben) [inline drivers/haptics/drv2605.c:608]: None of these properties exist in the binding, either due to bad names (start with `ti, ...` here but not in the yaml file), or not even defined at all in the yaml file (ex. `en-gpios`), so there is a lot of dead code and potentially a very stuck drv2605 when the EN input pin actually needs to be driven high to even be active :)

Note that i don't think we need/want `ti,` prefix in the first place

**Comment by rriveramcrus** [comment]
> @rriveramcrus any chance you could assemble a couple bullet points to add haptics to the release notes?

@kartben , Yes I can. (#81409)
